### PR TITLE
link: provide global const variables for section sizes

### PR DESCRIPTION
**Pull request by lorc**
Value of define VCORE_UNPG_RW_SZ is determined by linker script and
provided to C code as an symbol value. This is a standard way of sharing
linker variables with C code, which is described in ld manual.

Problem is that linker makes those symbols relocatable and ASLR code then
moves them to random places with rest of the OP-TEE image.

For example, on build for RCAR platform I am getting those entries in
relocation section:

[...]
000000004415b120 R_AARCH64_RELATIVE  *ABS*+0x0000000044100180
000000004415af60 R_AARCH64_RELATIVE  *ABS*+0x000000004415fc48
000000004415afb0 R_AARCH64_RELATIVE  *ABS*+0x00000000000a4000 <======
000000004415aef8 R_AARCH64_RELATIVE  *ABS*+0x000000004415c000
[...]

From programmer's point of view this looks like "constant" VCORE_UNPG_RW_SZ
has random value every boot.

Obvious approach is to provide section end address and then calculate size
on C side:

 #define ((size_t)(__vcore_unpg_rx_end - __vcore_unpg_rx_start))

But with this approach compiler can't determine value of VCORE_UNPG_RW_SZ
in compile time, so it can't initialize constant values in definitions like

 register_phys_mem_ul(MEM_AREA_TEE_RAM_RW, VCORE_UNPG_RW_PA,
					   VCORE_UNPG_RW_SZ);

from core_mmu.c

Basically, this leads to following constraints:

1. If we calculate section size in linker script, then compiler can use
   it as a compile-time constant, but this value will be mangled by ASLR
   in run-time.

2. We can't calculate section size in C code, because this value can't be
   used in compile-time.

This patch provides a workaround: section sizes are stored in constant
variables, which should be used in runtime, while original defines
should be used in compile-time only.

Fixes: 170e9084a84f ("core: add support for CFG_CORE_ASLR")
Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

---

It would be great to emit error if the define is used in run-time. But I have no idea how to implement that.

I tried to force linker to make mentioned symbols absolute, but with no success. Another approach is to make `gen_tee_bin.py` to omit those relocations. But I don't know reliable way to distinguish symbols with size values from all other symbols. We can try to employ simple heuristic: if symbol address is relatively small, then this is size constant. But I don't like this idea. 

And as always, I'm open for suggestions about naming :)

### CODE REVIEW

**Review by jforissier — COMMENTED** [review]
> 1. If we calculate section size in linker script, then compiler can use
>   it as a compile-time constant, but this value will be mangled by ASLR
>   in run-time.

There is something inconsistent in this explanation. If the size is calculated by the linker, then it is not available at compile time which happens before the link.

Are you saying that if we have:
```C
extern const size_t sz;
extern uint8_t start[];
extern uint8_t end[];
```
...then `register_phys_mem()` can take `sz` but not `(size_t)(start - end)`? What is the difference?

**Comment by jforissier** [comment]
> I tried to force linker to make mentioned symbols absolute, but with no success.

Have you tried adding the ABSOLUTE qualifier to the start and end symbols rather than their difference?
```
SECTIONS
{
    foo_start_abs = ABSOLUTE(.);
    /* some stuff ... */
    foo_end_abs = ABSOLUTE(.);
}

foo_sz = foo_end_abs - foo_start_abs;  # or possibly: foo_sz = ABSOLUTE(foo_end_abs - foo_start_abs);
```

**Comment by lorc** [comment]
> ...then `register_phys_mem()` can take `sz` but not `(size_t)(start - end)`? What is the difference?

Yes. This was the first thing I tried. 

`register_phys_mem()` is basically an initialization of constant structure. As compilation is done before linking, compiler can put a symbol reference as a value of any fields. But to do any computation, like subtraction of  two values, it needs to know exact values in compile time... And they are not available till linking stage.

**Comment by jforissier** [comment]
> > ...then `register_phys_mem()` can take `sz` but not `(size_t)(start - end)`? What is the difference?
> 
> Yes. This was the first thing I tried.
> 
> `register_phys_mem()` is basically an initialization of constant structure. As compilation is done before linking, compiler can put a symbol reference as a value of any fields. But to do any computation, like subtraction of two values, it needs to know exact values in compile time... And they are not available till linking stage.

Ah yes, obviously. Got it now. Perhaps `register_phys_mem()` could take start and end instead of size, although it would be a bit ugly.

**Comment by lorc** [comment]
> Have you tried adding the ABSOLUTE qualifier to the start and end symbols rather than their difference?

Yes, I did. It had no effect. According to documentation, absence of `ABSOLUTE` makes symbol value relative to section, if it is defined within section. But our `__vcore_unpg_rw_size` is already defined out of sections. So it is absolute by default. 

I also tried to put all those symbols in separate section with an fixed address `0x0`. But linker still generates relocations for them.

> Ah yes, obviously. Got it now. Perhaps `register_phys_mem()` could take start and end instead of size, although it would be a bit ugly.

Well, yes, this is another possibility. But this would require mega-patch to all platforms. And yes, most of calls to `register_phys_mem()` will become ugly.

File: core/arch/arm/kernel/sub.mk [hunk 1]
```diff
@@ -68,6 +68,7 @@ srcs-$(CFG_VIRTUALIZATION) += virtualization.c
 
 srcs-y += link_dummies_paged.c
 srcs-y += link_dummies_init.c
+srcs-y += link_symbols.c
```
[the remark this answers is not in the document]
Comment (lorc) [reply core/arch/arm/kernel/sub.mk:71]: Oops :-)
Now it should be fine.

**Comment by jforissier** [comment]
> > Have you tried adding the ABSOLUTE qualifier to the start and end symbols rather than their difference?
> 
> Yes, I did. It had no effect.

There has to be a way or else it is a linker bug IMO. How can I reproduce the issue on QEMU/QEMUv8?

**Comment by jenswi-linaro** [comment]
> There has to be a way or else it is a linker bug IMO. How can I reproduce the issue on QEMU/QEMUv8?

I suppose printing the value of one of the `_SZ` macros should be enough.

**Comment by jforissier** [comment]
> > There has to be a way or else it is a linker bug IMO. How can I reproduce the issue on QEMU/QEMUv8?
> 
> I suppose printing the value of one of the `_SZ` macros should be enough.

Indeed, `DMSG("VCORE_UNPG_RX_SZ=%p", (void *)VCORE_UNPG_RX_SZ);` prints a different (big) value every time. And yet the size is correct in the `add_phys_mem` boot message:
```
D/TC:0   add_phys_mem:557 VCORE_UNPG_RX_PA type TEE_RAM_RX 0x0e100000 size 0x0005e000
...
D/TC:0 0 boot_init_primary_late:1237 Executing at offset 0x9cd0c000 with virtual load address 0xaae0c000
...
D/TC:0 0 boot_init_primary_late:1251 VCORE_UNPG_RX_SZ=0x9cd6a000
```
I think the declaration:
```
core/arch/arm/include/kernel/linker.h:extern const uint8_t __vcore_unpg_ro_size[];
```
...is fundamentally wrong, because it says the symbol is an address and therefore subject to relocation, and the facts that things work in `register_phys_mem()` AFAICT is because dynamic relocations have not yet been applied. I understand we have to declare `__vcore_unpg_ro_size` like that because we get a compile error "initializer element is not constant" with the following declaration:
```
extern size_t __vcore_unpg_ro_size;
```
This is quite complex, misleading and error-prone IMO, but if we can't find a better way I think the appraoch in this PR is reasonable. I will suggest some changes to the comments.

**Review by jforissier — COMMENTED** [review]
OK, that's a tricky thing, here are my suggestions to make this as clear as possible...

File: core/arch/arm/kernel/link_symbols.c [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <kernel/linker.h>
+
+/*
+ * _R stands for Runtime. Those variables are used as workaround for a problem
+ * with linker emitting its symbols as relocatable ones. Size of each section is
+ * calculated by linker and presented to C code as a symbol. But linker marks
+ * those symbols as relocatable, so ASLR code randomizes their values each run.
+ * To mitigate this we can capture value of each symbol into a separate variable.
+ */
```
[the remark this answers is not in the document]
Comment (lorc) [reply core/arch/arm/kernel/link_symbols.c:14]: Yes, this is a tricky question indeed. Such things are not covered by standard and linker documentation does not give any guarantees.
Also, I can't imagine and compile-time or run-time check that will reveal any changes in compiler/linker behavior.

[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/link_symbols.c:14]: Runtime check should be possible if we define a linker symbol with a magic number and assign it to a global variable together with the rest. Later we can assert that we still have the magic number.

[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/link_symbols.c:14]: Yes or add `__vcore_unpg_rx_end` and then `assert(__vcore_unpg_rx_start + __vcore_unpg_rx_size_r == __vcore_unpg_rx_end)`?

[the remark this answers is not in the document]
Comment (lorc) [reply core/arch/arm/kernel/link_symbols.c:14]: This a great idea. I'll add the patch to this PR.

**Comment by lorc** [comment]
Well, I tried to add assert, as @jforissier suggested. 

Result is... well, it is expected now, when you see it. So, I did this in `link_symbols.c`:
```
	DMSG("__vcore_unpg_rw_end - __vcore_unpg_rw_start = %lx", __vcore_unpg_rw_end - __vcore_unpg_rw_start);
	DMSG("vcore_unpg_rw_size_r = %lx", vcore_unpg_rw_size_r);
	assert((size_t)(__vcore_unpg_rw_end - __vcore_unpg_rw_start) ==
	       vcore_unpg_rw_size_r);
```

And this is what I got:

```
D/TC:0 0 linker_check_size_for_relocs:30 __vcore_unpg_rw_end - __vcore_unpg_rw_start = a4000
D/TC:0 0 linker_check_size_for_relocs:31 vcore_unpg_rw_size_r = ffffffffc04f6000
E/TC:0 0 assertion '(size_t)(__vcore_unpg_rw_end - __vcore_unpg_rw_start) == vcore_unpg_rw_size_r' failed at core/arch/arm/kernel/link_symbols.c:32 <linker_check_size_for_relocs>
```
```
D/TC:0 0 linker_check_size_for_relocs:30 __vcore_unpg_rw_end - __vcore_unpg_rw_start = a4000                                                                                                   
D/TC:0 0 linker_check_size_for_relocs:31 vcore_unpg_rw_size_r = fffffffff65c3000                                                                                                               
E/TC:0 0 assertion '(size_t)(__vcore_unpg_rw_end - __vcore_unpg_rw_start) == vcore_unpg_rw_size_r' failed at core/arch/arm/kernel/link_symbols.c:32 <linker_check_size_for_relocs>
E
```

Note that value of `vcore_unpg_rw_size_r` is different every time. Apparently, compiler decided to optimize the code and replaced access to `vcore_unpg_rw_size_r` with direct access to `(size_t)__vcore_unpg_rw_size`.

So, I'm going to rework the PR by assigning all `_size_r` variables like this:

```
 const size_t vcore_unpg_rw_size_r = (size_t)(__vcore_unpg_rw_end - __vcore_unpg_rw_start);
```

This will fix the issue once and for all. And we will not need the second patch with `assert`.

**Comment by jforissier** [comment]
The assert triggers in the QEMU 32-bit build :-/ I checked with GDB, and also with a script I wrote to help parse relocations, and clearly `vcore_unpg_rw_size_r` has been relocated:
```
$ ~/bin/print_relocations.py out/arm/core/tee.elf | grep vcore_unpg_rw_size_r
0xe172030 in .data <vcore_unpg_rw_size_r> REL type R_ARM_RELATIVE sym 0
```

The solution is perhaps to drop the idea of having global constants and define some `_R` macros in `linker.h` instead, like so:
```diff
 #define VCORE_UNPG_RW_SZ       ((size_t)__vcore_unpg_rw_size)
+#define VCORE_UNPG_RW_SZ_R     ((size_t)(__vcore_unpg_rw_end - \
+                                         __vcore_unpg_rw_start))
 #define VCORE_NEX_RW_PA        ((unsigned long)__vcore_nex_rw_start)
```
(and adding the required _end symbols to `kern.ld.s` of course).

**Comment by jenswi-linaro** [comment]
> The solution is perhaps to drop the idea of having global constants and define some _R macros in linker.h instead

+1 to that.

**Comment by lorc** [comment]
> The solution is perhaps to drop the idea of having global constants and define some `_R` macros in `linker.h` instead, like so:

Yes, it seems, this is the best way. Had to use it from the beginning...

Anyways, I reworked the patch.

File: core/arch/arm/include/kernel/linker.h [hunk 3]
```diff
@@ -22,33 +22,59 @@ extern const uint8_t __exidx_end[];
...
+ * mostly here for use in register_phys_mem() where the _SZ_R macros would be
+ * rejected by the compiler ('initializer element is not constant'). Other uses
+ * are likely to be incorrect.
+ * Reiterating, _SZ_C defines are for compile-time definitions, when constant
+ * values are required and _SC_R are for run-time uses.
+ */
+
 #define VCORE_UNPG_RX_PA	((unsigned long)__vcore_unpg_rx_start)
-#define VCORE_UNPG_RX_SZ	((size_t)__vcore_unpg_rx_size)
+#define VCORE_UNPG_RX_SZ_C	((size_t)__vcore_unpg_rx_size)
```
Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/linker.h:39]: To discourage the use of the `_C` defines perhaps they should be having a `_UNSAFE` suffix instead.
We could then drop the `_R` suffix.

File: core/arch/arm/kernel/kern.ld.S [hunk 4]
```diff
@@ -444,14 +444,19 @@ __vcore_unpg_rx_size = __flatmap_unpg_rx_size + __flatmap_unpg_ro_size;
 __vcore_unpg_ro_size = 0;
 #endif
 
+__vcore_unpg_rx_end = __vcore_unpg_rx_start + __vcore_unpg_rx_size;
```
Comment (jforissier) [inline core/arch/arm/kernel/kern.ld.S:447]: Minor: remove blank line above for consistency with `_start` and `_size`

File: core/arch/arm/kernel/link_dummy.ld [hunk 5]
```diff
@@ -75,16 +75,22 @@ __tmp_hashes_size = .;
...
+__vcore_nex_rw_end = .;
 __vcore_unpg_ro_size = .;
 __vcore_unpg_ro_start = .;
+__vcore_unpg_ro_end = .;
 __vcore_unpg_rw_size = .;
 __vcore_unpg_rw_start = .;
+__vcore_unpg_rw_end = .;
 __vcore_unpg_rx_size = .;
 __vcore_unpg_rx_start = .;
+__vcore_unpg_rx_end = .;
```
Comment (jforissier) [inline core/arch/arm/kernel/link_dummy.ld:93]: Sort alphabetically

**Comment by jforissier** [comment]
LGTM. I suggest some changes in the commit text.

```
Value of define VCORE_UNPG_RW_SZ is determined by linker script and
provided to C code as an symbol value. This is a standard way of sharing
```
"as **a** symbol value **(__vcore_unpg_rw_size)**".

```
linker variables with C code, which is described in ld manual.

Problem is that linker makes those symbols relocatable and ASLR code then
```
"the linker **sometimes** makes"
```
moves them to random places with rest of the OP-TEE image.

For example, on build for RCAR platform I am getting those entries in
relocation section:

[...]
000000004415b120 R_AARCH64_RELATIVE  *ABS*+0x0000000044100180
000000004415af60 R_AARCH64_RELATIVE  *ABS*+0x000000004415fc48
000000004415afb0 R_AARCH64_RELATIVE  *ABS*+0x00000000000a4000 <======
000000004415aef8 R_AARCH64_RELATIVE  *ABS*+0x000000004415c000
[...]

From programmer's point of view this looks like "constant" VCORE_UNPG_RW_SZ
has random value every boot.

Obvious approach is to provide section end address and then calculate size
on C side:

 #define ((size_t)(__vcore_unpg_rx_end - __vcore_unpg_rx_start))
```
missing **VCORE_UNPG_RW_SZ**
```
But with this approach compiler can't determine value of VCORE_UNPG_RW_SZ
in compile time, so it can't initialize constant values in definitions like
```
Sorry for being a bit pedantic here, but we should try to be precise because things can get confusing for anyone trying to understand what is really happening and referencing the C standard.

I think you should drop "can't determine value of VCORE_UNPG_RW_SZ in compile time" because strictly speaking the value is not known at compile time, I mean link time is involved, and the important thing is the end of the sentence anyways.

The thing we are talking about here is a *constant expression*, that is the name used in the C standard, which has a precise definition which is sometimes counter-intuitive. For instance if you have `const int N = 1;` then `N` is *not* a constant expression although it is constant (never changes) and known at compile-time...

```
 register_phys_mem_ul(MEM_AREA_TEE_RAM_RW, VCORE_UNPG_RW_PA,
					   VCORE_UNPG_RW_SZ);

from core_mmu.c

Basically, this leads to following constraints:

1. If we calculate section size in linker script, then compiler can use
   it as a compile-time constant, but this value will be mangled by ASLR
```
"as a **constant expression**", "**may** be mangled" (doesn't always happen, Clang 11 for example)
```
   in run-time.
```
"**at** run time"
```
2. We can't calculate section size in C code, because this value can't be
   used in compile-time.
```
"used **as a constant expression**."
```

This patch provides a workaround around this issue by providing two sets of
definitions: old _SZ definition is renamed to _SZ_UNSAFE and it should be
used only in compile-time, while new _SZ definition is for use in run-time.
```
"...used only in places where a constant expression is required and provided it is referenced only before dynamic relocations have been applied, while the new _SZ definition can be used in all other situations."
```
Value of _new SZ is obtained by deducting section start address from end
address. Additional linker symbols were introduced to provide section end
```
"**are** introduced"
```
addresses.

Fixes: 170e908 ("core: add support for CFG_CORE_ASLR")
Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
```

With that:

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

Thanks for spending time on fixing this!

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by lorc** [comment]
@jforissier, thank you for help with the commit description. I updated it and added tags.

**Pull request merged**